### DISCUSSION TITLE
Magma quick start build error at "make run"

### DISCUSSION THREAD

**Discussion by jni2000** [magma/magma#14598] [Q&A]
Hello everyone. I have followed https://magma.github.io/magma/docs/basics/prerequisites
and now is trying to go through: https://magma.github.io/magma/docs/basics/quick_start_guide. After resolving pip package conflicts with regards to protobuf and grpcio-tools etc., I was able to move on in "make run" stage running in magma virtualbox but blocked at a build error as below. Can anyone please help with any clues ?

**## a warning here -- I guess not critical**
/home/vagrant/build/python/bin/pip3 install -q -U --cache-dir ~/.pipcache "pip>=20.3.2"
Installing egg link for lte
/home/vagrant/build/python/bin/pip3 install -q -U --cache-dir ~/.pipcache --no-build-isolation -e .[dev]
**WARNING: No metadata found in /usr/local/lib/python3.8/dist-packages**
make[2]: Leaving directory '/home/vagrant/magma/lte/gateway/python'

**## passed a lot of lines and then see the following error**
[21/24] Performing configure step for 'MagmaCore'
-- The C compiler identification is GNU 9.3.0
-- The CXX compiler identification is GNU 9.3.0
-- Check for working C compiler: /usr/lib/ccache/cc
-- Check for working C compiler: /usr/lib/ccache/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/lib/ccache/c++
-- Check for working CXX compiler: /usr/lib/ccache/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Protobuf: /usr/lib/x86_64-linux-gnu/libprotobuf.so;-lpthread (found version "3.6.1") 
Build type is Debug
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.1") 
-- Found Threads: TRUE  
-- Checking for one of the modules 'libconfig'
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Checking for one of the modules 'openssl'
-- Checking for one of the modules 'libcrypto'
-- Checking for one of the modules 'nettle'
Proto_srcs are /home/vagrant/build/c/core/oai/common/lte/protos/oai/common_types.pb.cc;/home/vagrant/build/c/core/oai/common/orc8r/protos/common.pb.cc;/home/vagrant/build/c/core/oai/common/orc8r/protos/redis.pb.cc;/home/vagrant/build/c/core/oai/common/lte/protos/mconfig/mconfigs.pb.cc
os_nameubuntu
Proto_srcs are /home/vagrant/build/c/core/oai/lib/sgs_client/feg/protos/csfb.pb.cc;/home/vagrant/build/c/core/oai/lib/sgs_client/feg/protos/csfb.grpc.pb.cc;/home/vagrant/build/c/core/oai/lib/sgs_client/lte/protos/mconfig/mconfigs.pb.cc
Proto_srcs are /home/vagrant/build/c/core/oai/lib/sgs_client/feg/protos/csfb.pb.cc;/home/vagrant/build/c/core/oai/lib/sgs_client/feg/protos/csfb.grpc.pb.cc;/home/vagrant/build/c/core/oai/lib/sgs_client/lte/protos/mconfig/mconfigs.pb.cc;/home/vagrant/build/c/core/oai/lib/sgs_client/orc8r/protos/common.pb.cc
Proto_srcs are /home/vagrant/build/c/core/oai/lib/sms_orc8r_client/lte/protos/sms_orc8r.pb.cc;/home/vagrant/build/c/core/oai/lib/sms_orc8r_client/lte/protos/sms_orc8r.grpc.pb.cc;/home/vagrant/build/c/core/oai/lib/sms_orc8r_client/lte/protos/mconfig/mconfigs.pb.cc
-- Checking for one of the modules 'libxml-2.0'
Proto_srcs are /home/vagrant/build/c/core/oai/tasks/nas/lte/protos/oai/nas_state.pb.cc;/home/vagrant/build/c/core/oai/tasks/nas/lte/protos/oai/spgw_state.pb.cc;/home/vagrant/build/c/core/oai/tasks/nas/lte/protos/oai/common_types.pb.cc;/home/vagrant/build/c/core/oai/tasks/nas/lte/protos/oai/std_3gpp_types.pb.cc
Generated /home/vagrant/build/c/core/oai/tasks/s1ap/r15/Makefile.am.libasncodec
Generated S1ap_asn_constant.h
Generated /home/vagrant/build/c/core/oai/tasks/ngap/r16/Makefile.am.libasncodec
Generated pdu_collection.c
Generated /home/vagrant/build/c/core/oai/tasks/ngap/r16/converter-example.mk
Generated pdu_collection.c
Generated /home/vagrant/build/c/core/oai/tasks/ngap/r16/Makefile.am.asn1convert
Generated Ngap_asn_constant.h
s8 Proto_srcs are /home/vagrant/build/c/core/oai/tasks/sgw_s8/lte/protos/oai/spgw_state.pb.cc;/home/vagrant/build/c/core/oai/tasks/sgw_s8/lte/protos/oai/sgw_state.pb.cc;/home/vagrant/build/c/core/oai/tasks/sgw_s8/lte/protos/oai/common_types.pb.cc;/home/vagrant/build/c/core/oai/tasks/sgw_s8/lte/protos/oai/std_3gpp_types.pb.cc
s8 Proto_includes are /home/vagrant/build/c/core/oai/tasks/sgw_s8/lte/protos/oai/spgw_state.pb.h;/home/vagrant/build/c/core/oai/tasks/sgw_s8/lte/protos/oai/sgw_state.pb.h;/home/vagrant/build/c/core/oai/tasks/sgw_s8/lte/protos/oai/common_types.pb.h;/home/vagrant/build/c/core/oai/tasks/sgw_s8/lte/protos/oai/std_3gpp_types.pb.h
-- Configuring done
-- Generating done
-- Build files have been written to: /home/vagrant/build/c/core/oai
[22/24] Performing build step for 'MagmaCore'
**FAILED: MagmaCore-prefix/src/MagmaCore-stamp/MagmaCore-build** 

**## passed a lot of others and then finally see**
ninja: build stopped: subcommand failed.
ninja: build stopped: subcommand failed.
make: *** [Makefile:161: build_oai] Error 1

**Comment by jni2000** [comment]
Additional error information from /build/c/core/oai/CMakeFiles/CMakeError.log

Performing C SOURCE FILE Test CMAKE_HAVE_LIBC_PTHREAD failed with the following output:
Change Dir: /home/vagrant/build/c/core/oai/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/ninja cmTC_5b5a1 && [1/2] Building C object CMakeFiles/cmTC_5b5a1.dir/src.c.o
[2/2] Linking C executable cmTC_5b5a1
FAILED: cmTC_5b5a1 
: && /usr/lib/ccache/cc -DCMAKE_HAVE_LIBC_PTHREAD   CMakeFiles/cmTC_5b5a1.dir/src.c.o  -o cmTC_5b5a1   && :
/usr/bin/ld: CMakeFiles/cmTC_5b5a1.dir/src.c.o: in function `main':
src.c:(.text+0x46): undefined reference to `pthread_create'
/usr/bin/ld: src.c:(.text+0x52): undefined reference to `pthread_detach'
/usr/bin/ld: src.c:(.text+0x63): undefined reference to `pthread_join'
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.


Source file was:
#include <pthread.h>

void* test_func(void* data)
{
  return data;
}

int main(void)
{
  pthread_t thread;
  pthread_create(&thread, NULL, test_func, NULL);
  pthread_detach(thread);
  pthread_join(thread, NULL);
  pthread_atfork(NULL, NULL, NULL);
  pthread_exit(NULL);

  return 0;
}

Determining if the function pthread_create exists in the pthreads failed with the following output:
Change Dir: /home/vagrant/build/c/core/oai/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/ninja cmTC_b7d06 && [1/2] Building C object CMakeFiles/cmTC_b7d06.dir/CheckFunctionExists.c.o
[2/2] Linking C executable cmTC_b7d06
FAILED: cmTC_b7d06 
: && /usr/lib/ccache/cc -DCHECK_FUNCTION_EXISTS=pthread_create   CMakeFiles/cmTC_b7d06.dir/CheckFunctionExists.c.o  -o cmTC_b7d06  -lpthreads && :
/usr/bin/ld: cannot find -lpthreads
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed

**Comment by mpfirrmann** [comment]
Which branch are you on? I can't reproduce the error. Building the AGW in the development environment as described in the Quick Start Guide should not throw an error.

**Reply by jni2000** [reply]
Thanks, Marco Pfirrmann, for replying. I was following the Quick Start Guide and building magma 1.8.

Here are some progress I made in the past a few days.
1. After downgrading virtualbox from version 7.0 to version 6.1 (which I believe was the version used when magma 1.8 was released), the above errors were not seen any more.
2. However, I now see the following errors when running "make run" in magma virtualbox.
`/home/vagrant/magma/lte/gateway/c/core/oai/lib/openflow/controller/GTPApplication.cpp:415:13: error: expected type-specifier
  415 |         new of13::TunnelIPv6Dst(remote_ipv6.s6_addr));
      |             ^~~~
[2004/2088] Building CXX object lib/pipelined_client/CMakeFiles/LIB_PIPELINED_CLIENT.dir/proto_converters.cpp.o
[2005/2088] Building CXX object lib/pipelined_client/CMakeFiles/LIB_PIPELINED_CLIENT.dir/PipelinedClientAPI.cpp.o
[2006/2088] Building CXX object lib/pipelined_client/CMakeFiles/LIB_PIPELINED_CLIENT.dir/lte/protos/mobilityd.pb.cc.o
[2007/2088] Building CXX object lib/pipelined_client/CMakeFiles/LIB_PIPELINED_CLIENT.dir/PipelinedServiceClient.cpp.o
[2008/2088] Building CXX object lib/pipelined_client/CMakeFiles/LIB_PIPELINED_CLIENT.dir/lte/protos/pipelined.pb.cc.o
ninja: build stopped: subcommand failed.
ninja: build stopped: subcommand failed.
make: *** [Makefile:161: build_oai] Error 1
vagrant@magma-dev-focal:~/magma/lte/gateway$ 

`
Any clue what I was missing ?

**Reply by swapniil** [reply]
I'm facing same problem while building magma

The thing is that, I can see the `class TunnelIPv4Dst` is declared in the `fluid` headers.
The header for `TunnelIPv6Dst` is missing and so the compilation fails.

```
vagrant@magma-dev-focal:~/magma/lte$ grep TunnelIPv4Dst /usr/include/fluid/ -rn
/usr/include/fluid/of13/of13match.hh:1118:class TunnelIPv4Dst: public OXMTLV {
/usr/include/fluid/of13/of13match.hh:1123:    TunnelIPv4Dst();
/usr/include/fluid/of13/of13match.hh:1124:    TunnelIPv4Dst(IPAddress value);
/usr/include/fluid/of13/of13match.hh:1125:    TunnelIPv4Dst(IPAddress value, IPAddress mask);
/usr/include/fluid/of13/of13match.hh:1126:    ~TunnelIPv4Dst() {
/usr/include/fluid/of13/of13match.hh:1130:    virtual TunnelIPv4Dst* clone() const {
/usr/include/fluid/of13/of13match.hh:1131:        return new TunnelIPv4Dst(*this);
vagrant@magma-dev-focal:~/magma/lte$ 
vagrant@magma-dev-focal:~/magma/lte$ 
vagrant@magma-dev-focal:~/magma/lte$ 
vagrant@magma-dev-focal:~/magma/lte$ grep TunnelIPv6Dst /usr/include/fluid/ -rn
vagrant@magma-dev-focal:~/magma/lte$ 

```

Based on the contents of package the library code is not available to be modified.

```
vagrant@magma-dev-focal:~/magma$ dpkg -L magma-libfluid
/.
/usr
/usr/lib
/usr/lib/libfluid_msg.so.0.0.0
/usr/lib/libfluid_msg.la
/usr/lib/libfluid_base.a
/usr/lib/pkgconfig
/usr/lib/pkgconfig/libfluid_base.pc
/usr/lib/pkgconfig/libfluid_msg.pc
/usr/lib/libfluid_msg.a
/usr/lib/libfluid_base.la
/usr/lib/libfluid_base.so.0.0.0
/usr/share
/usr/share/doc
/usr/share/doc/magma-libfluid
/usr/share/doc/magma-libfluid/changelog.gz
/usr/include
/usr/include/fluid
/usr/include/fluid/util
/usr/include/fluid/util/util.h
/usr/include/fluid/util/ipaddr.hh
/usr/include/fluid/util/ethaddr.hh
/usr/include/fluid/OFServer.hh
/usr/include/fluid/of13
/usr/include/fluid/of13/of13action.hh
/usr/include/fluid/of13/of13common.hh
/usr/include/fluid/of13/of13instruction.hh
/usr/include/fluid/of13/of13meter.hh
/usr/include/fluid/of13/openflow-13.h
/usr/include/fluid/of13/of13match.hh
/usr/include/fluid/OFConnection.hh
/usr/include/fluid/of13msg.hh
/usr/include/fluid/OFClient.hh
/usr/include/fluid/ofcommon
/usr/include/fluid/ofcommon/openflow-common.hh
/usr/include/fluid/ofcommon/action.hh
/usr/include/fluid/ofcommon/common.hh
/usr/include/fluid/ofcommon/msg.hh
/usr/include/fluid/of10
/usr/include/fluid/of10/of10action.hh
/usr/include/fluid/of10/of10match.hh
/usr/include/fluid/of10/of10common.hh
/usr/include/fluid/of10/openflow-10.h
/usr/include/fluid/OFServerSettings.hh
/usr/include/fluid/of10msg.hh
/usr/include/fluid/base
/usr/include/fluid/base/BaseOFConnection.hh
/usr/include/fluid/base/BaseOFServer.hh
/usr/include/fluid/base/BaseOFClient.hh
/usr/include/fluid/base/EventLoop.hh
/usr/include/fluid/TLS.hh
/usr/lib/libfluid_base.so
/usr/lib/libfluid_msg.so.0
/usr/lib/libfluid_base.so.0
/usr/lib/libfluid_msg.so
vagrant@magma-dev-focal:~/magma$ 
```


There is this file in magma repository `/home/vagrant/magma/third_party/build/patches/libfluid/libfluid_msg_patches/0004-Add-TunnelIPv6Dst-support.patch` to enable IPv6.
I'm not able to find the code so not able complete it.

**Reply by swapniil** [reply]
The magma code needs the `0.1.0.7` version of the `magma-libfluid`.

I see that the version `magma-libfluid/focal-ci 0.1.0.7-1 amd64` of `magma-libfluid`  is not available at artifact server.

```
root@magma-dev-focal:/home/vagrant/magma/lte/gateway# apt install magma-libfluid
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following packages were automatically installed and are no longer required:
  libnetplan0 linux-headers-5.4.0-73 linux-headers-5.4.0-73-generic linux-image-5.4.0-73-generic linux-modules-5.4.0-73-generic
Use 'sudo apt autoremove' to remove them.
The following packages will be upgraded:
  magma-libfluid
1 upgraded, 0 newly installed, 0 to remove and 124 not upgraded.
Need to get 5375 kB of archives.
After this operation, 42.0 kB of additional disk space will be used.
Err:1 https://artifactory.magmacore.org/artifactory/debian-test focal-ci/main amd64 magma-libfluid amd64 0.1.0.7-1
  Could not connect to artifactory.magmacore.org:443 (51.178.63.140). - connect (111: Connection refused)
E: Failed to fetch https://artifactory.magmacore.org/artifactory/debian-test/pool/focal/magma-libfluid_0.1.0.7-1_amd64.deb  Could not connect to artifactory.magmacore.org:443 (51.178.63.140). - connect (111: Connection refused)
E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
root@magma-dev-focal:/home/vagrant/magma/lte/gateway# 
```

If you get this package then probably your build might get be succeed. At least you'll have a fix for this issue.

If not then comment the else block on `/home/vagrant/magma/lte/gateway/c/core/oai/lib/openflow/controller/GTPApplication.cpp:415; line

**Reply by LKreutzer** [reply]
Hey, are you using the current artifactory? I think the current URL is `https://linuxfoundation.jfrog.io/artifactory`

**Reply by swapniil** [reply]
I'm not modifying anything in tag v1.8.0. Its whatever there as is I'm using. Can you tell me in which file I can change this?

**Reply by swapniil** [reply]
```
vagrant@magma-dev-focal:~/magma/lte/gateway$ sudo apt update 
Hit:1 http://security.ubuntu.com/ubuntu focal-security InRelease
Hit:2 http://archive.ubuntu.com/ubuntu focal InRelease                          
Hit:3 http://archive.ubuntu.com/ubuntu focal-updates InRelease                  
Hit:4 http://archive.ubuntu.com/ubuntu focal-backports InRelease               
Ign:5 https://linuxfoundation.jfrog.io/artifactory focal-ci InRelease          
Err:6 https://linuxfoundation.jfrog.io/artifactory focal-ci Release
  404   [IP: 52.41.191.44 443]
Reading package lists... Done
E: The repository 'https://linuxfoundation.jfrog.io/artifactory focal-ci Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
vagrant@magma-dev-focal:~/magma/lte/gateway$ sudo apt search magma-libfluid
```

The https://linuxfoundation.jfrog.io/artifactory is also down it looks.

**Comment by LKreutzer** [comment]
Regarding the original error, this may be solved by following https://github.com/magma/magma/issues/12362#issuecomment-1430945413
